File: Backend/src/spe/pipeline/operators/base/sources/randomData.py
import json
import random
import string
import time

from spe.pipeline.operators.source import Source


class RandomData(Source):
    def __init__(self,  opID: int):
        super(RandomData, self).__init__(opID, 0, 1)

        self.rate = 0
        self.limitRate = True

    def setData(self, data: json):
        self.rate = data["rate"]
        self.limitRate = data["limitRate"]

    def getData(self) -> dict:
        return {"rate": self.rate, "limitRate": self.limitRate}

    def _runSource(self):
        while self.isRunning():
            data = {
                "val1": random.random(),
                "val2": random.randint(1, 100),
                "val3": "".join(random.choices(string.ascii_letters, k=10)),
                "val4": [random.random() for _ in range(10)]
            }

            self._produce((data,))

            if self.limitRate:
                sleepDuration = 1 / self.rate

                if sleepDuration > 1e-3:
                    time.sleep(1 / self.rate)

File: frontend/src/scripts/pipeline/operators/modules/base/sources/RandomData.js
import {Definition} from "@/scripts/pipeline/operators/Definition";
import {NumberParam} from "@/scripts/pipeline/operators/modules/base/params/NumberParam";
import {BoolParam} from "@/scripts/pipeline/operators/modules/base/params/BoolParam";
import {anySocket} from "@/scripts/pipeline/operators/modules";
import {SocketDef} from "@/scripts/pipeline/SvSocket";

export default class _RandomData extends Definition {
    constructor(pathIdentifier){
        super("RandomData", "Random Data", pathIdentifier,
            "Emits randomly generated data containing a dictionary of {val1: float, val2: int, val3: string, val4: list[float]} for testing purposes.", true);
    }

    build(operator) {
        let limitRate = new BoolParam("limitRate", true, "Limit Rate",
            "If the source should produce tuples in a fixed rate");
        let rate = new NumberParam("rate", 30, 0, null,
            "Rate", "How many data tuples per second are generated.");

        limitRate.onChangeCallback = (val) => { rate.show = val; };

        this._construct(operator,
            [],
            [new SocketDef(anySocket, "Data")],
            [limitRate, rate]);
    }
}
